lib/connection.js

```diff
@@ -84,784 +83,749 @@ function Connection(r, options, resolve, reject) {
-        reject(new Err.ReqlDriverError('Could not derive the key. Error:' + error.toString()).setOperational());
+        throw new Err.ReqlDriverError('Could not derive the key. Error:' + error.toString()).setOperational()
       }
-      CACHE_PBKDF2[cacheKey] = saltedPassword;
-      helper.tryCatch(function() {
-        self._sendProof(messageServer.authentication, randomNonce, saltedPassword);
-      }, function(err) {
-        // The TCP connection is open, but the ReQL connection wasn't established.
-        // We can just abort the whole thing
-        self.open = false;
-        reject(new Err.ReqlDriverError('Failed to perform handshake with '+self.host+':'+self.port).setOperational());
-      });
+      CACHE_PBKDF2[cacheKey] = saltedPassword
+      self._sendProof(handshake.authentication, randomNonce, saltedPassword)
     })
   }
 }
 
-Connection.prototype._sendProof = function(authentication, randomNonce, saltedPassword) {
-  var clientFinalMessageWithoutProof = "c=biws,r=" + randomNonce;
-  var clientKey = crypto.createHmac("sha256", saltedPassword).update("Client Key").digest()
-  var storedKey = crypto.createHash("sha256").update(clientKey).digest()
+Connection.prototype._sendProof = function (authentication, randomNonce, saltedPassword) {
+  var clientFinalMessageWithoutProof = 'c=biws,r=' + randomNonce
+  var clientKey = crypto.createHmac('sha256', saltedPassword).update('Client Key').digest()
+  var storedKey = crypto.createHash('sha256').update(clientKey).digest()
 
   var authMessage =
-      "n=" + this.user + ",r=" + this.randomString + "," +
-      authentication + "," +
+      'n=' + this.user + ',r=' + this.randomString + ',' +
+      authentication + ',' +
       clientFinalMessageWithoutProof
 
-  var clientSignature = crypto.createHmac("sha256", storedKey).update(authMessage).digest()
+  var clientSignature = crypto.createHmac('sha256', storedKey).update(authMessage).digest()
   var clientProof = helper.xorBuffer(clientKey, clientSignature)
 
-  var serverKey = crypto.createHmac("sha256", saltedPassword).update("Server Key").digest()
-  this.serverSignature = crypto.createHmac("sha256", serverKey).update(authMessage).digest()
+  var serverKey = crypto.createHmac('sha256', saltedPassword).update('Server Key').digest()
+  this.serverSignature = crypto.createHmac('sha256', serverKey).update(authMessage).digest()
 
-  this.state = 2
   var message = JSON.stringify({
-    authentication: clientFinalMessageWithoutProof + ",p=" + clientProof.toString("base64")
+    authentication: clientFinalMessageWithoutProof + ',p=' + clientProof.toString('base64')
   })
-  this.connection.write(Buffer.concat([new Buffer(message.toString()), NULL_BUFFER]))
+
+  try {
+    this.connection.write(Buffer.concat([Buffer.from(message.toString()), NULL_BUFFER]))
+  } catch (err) {
+    // The TCP connection is open, but the ReQL connection wasn't established.
+    // We can just abort the whole thing
+    // self.open = false;
```

Comment:
This code path is only reached `if (self.open === false)` so i think this is redundant.